### ISSUE TITLE
Is there a way to  access x-amz-meta fields (user Metadata) on an S3 object?

### ISSUE THREAD

**Issue by orbweaver** [pplu/aws-sdk-perl#131]
Or is there still work that needs to be done?

**Comment by orbweaver** [comment]
UPDATE: I was able to pull the metadata field my modifying GetObjectOutput.pm adding a line for the specific metadata field (this example is called DES).

has DES => (is => 'ro', isa => 'Str', traits => ['ParamInHeader'], header_name => 'x-amz-meta-des');

I was able to place the metatadata into the upload by putting same line in PutObject.pm.
to extract the metadata: 
$res = $s3->GetObject(%parameters);
my $DESinfo = $res->{'DES'};

so.....it seems I'm close but it requires manually adding a line for each piece of metadata in GetObjectOutput and PutObject.

**Comment by pplu** [comment]
Great that you found the way to "add missing attributes" to a class :)

Can you point me to where the 'x-amz-meta-des' header is documented?

Note: I've observed that you are accessing `$res->{'DES'}`. $res is really an object, so the correct way to access the value would be `$res->DES`. This is valid for all objects returned by Paws (they are Moose objects, and each attribute is accessed as a method, not a hash reference, which is "reaching into it's internals". I strongly advise you don't access $res as a hash reference).

**Comment by Grinnz** [comment]
This header seems to be documented here: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html#RESTObjectPUT-requests and here: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html#RESTObjectGET-responses

**Comment by pplu** [comment]
Hi,

There is a problem with the meta keys: They are arbitrary, and can be set by the user, so I think they should be accessible through an attribute `->Meta` that is a Map type (that holds all the Key/Value pairs for meta headers.

**Comment by orbweaver** [comment]
yes.  as described above, the only way I can set or get any meta data is to modify GetObjectOutput.pm and PutObject.pm to include a 'ParamInHeader', like this:
has DES => (is => 'ro', isa => 'Str', traits => ['ParamInHeader'], header_name => 'x-amz-meta-DES' );

I'm not sure how ->Meta would work, but I love the idea of getting/setting key and value pairs for meta headers.

**Comment by glenveegee** [comment]
Unless I'm doing something wrong, there does appear to be a bug here. I've tried to add metadata

<pre>
$response = $s3->PutObject(
          Bucket   => $bucketname,
          Key      => 'uploads/test/foo.txt',
          Body     => 'foobar',
          Metadata => {
            DES => q[des_value],
            ted => q[ted_value],
            Sue => q[sue_data]
          }
      );
</pre>

And also fetch Metadata from objects with metadata already set

<pre>
    $response = $s3->GetObject(
        Bucket => $bucketname,
        Key    => 'uploads/test/foo.txt'
    );
</pre>
 

PutObject throws an exception while GetObject returns an object without Metadata set

<pre>
          'success' => 1,
          'reason' => 'OK',
          'url' => 'https://s3-eu-west-1.amazonaws.com/bucket/uploads/test/foo.txt',
          'status' => '200',
          'headers' => {
                         'x-amz-request-id' => '6F86841499635A29',
                         'accept-ranges' => 'bytes',
                         'etag' => '"asdfasdfasdfasdfasdfasdf"',
                         'connection' => 'close',
                         'content-length' => '6',
                         'last-modified' => 'Thu, 01 Dec 2016 11:02:32 GMT',
                         'x-amz-meta-foo' => 'bar',
                         'x-amz-meta-roo' => 'car',
                         'server' => 'AmazonS3',
                         'content-type' => 'application/octet-stream',
                         'x-amz-meta-noo' => 'tar',
                         'date' => 'Thu, 01 Dec 2016 12:00:40 GMT'
                       },
          'content' => 'foobar'
        };
 at aws-sdk-perl/lib/Paws/Net/RetryCallerRole.pm line 18.
$VAR1 = bless( {
                 'Body' => 'foobar',
                 'ContentLength' => '6',
                 'AcceptRanges' => 'bytes',
                 'LastModified' => 'Thu, 01 Dec 2016 11:02:32 GMT',
                 '_request_id' => '6F86841499635A29',
                 'ContentType' => 'application/octet-stream',
                 'ETag' => '"asdfasdfasdfasdfasdfasdf"'
               }, 'Paws::S3::GetObjectOutput' );
</pre>

**Comment by PopeFelix** [comment]
@glenveegee , is the error you are getting similar to: "Can't call method "meta" on unblessed reference at /Users/popefelix/.perlbrew/libs/perl-5.16.3@p5-bean-datastore/lib/perl5/Paws/Net/RestXmlCaller.pm line 23."

If so, I can tell you why this error is happening, but I can't tell you how to fix it. :( 

See lib/Paws/Net/RestXmlCaller.pm line 46:

`my %complex_value = $self->_to_querycaller_params($params->$att);`

This calls `_to_querycaller_params`, which expects a Moose object as its argument and iterates over its attributes. In the case of attributes with a simple type constraint (i.e. a scalar value), this is fine. However, the `Map` attribute of `Paws::Metadata` has a HashRef type constraint, which gets handled recursively, and `_to_querycaller_params` does not correctly handle bare hashrefs. 

Long story short, the fix is to handle hashrefs, but I don't know how to do that properly.

**Comment by PopeFelix** [comment]
So if you look at https://github.com/PopeFelix/aws-sdk-perl/tree/fix-s3-metadata, there's a fix in there, but a lot of the S3 metadata stuff is hardcoded

**Comment by pplu** [comment]
Version 0.32 with support for this has hit the CPAN
